devnet-sdk/system/periphery/go-ethereum/fees.go

```diff
@@ -0,0 +1,81 @@
+package goethereum
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+var (
+	// Ensure that the feeEstimator implements the FeeEstimator interface
+	_ FeeEstimator = (*eip1559eeEstimator)(nil)
+)
+
+// FeeEstimator is a generic fee estimation interface (not specific to EIP-1559)
+type FeeEstimator interface {
+	EstimateFees(ctx context.Context, opts *bind.TransactOpts) (*bind.TransactOpts, error)
+}
+
+// eip1559eeEstimator is a fee estimator that uses EIP-1559 fee estimation
+type eip1559eeEstimator struct {
+	// Access to the Ethereum client is needed to get the fee information from the chain
+	client FeeEthClient
+
+	// The tip multiplier is used to increase the maxPriorityFeePerGas (GasTipCap) by a factor
+	tipMultiplier *big.Int
+}
+
+func NewEIP1559FeeEstimator(client FeeEthClient) *eip1559eeEstimator {
+	return &eip1559eeEstimator{
+		client:        client,
+		tipMultiplier: big.NewInt(1),
+	}
+}
+
+func (f *eip1559eeEstimator) WithTipMultiplier(multiplier *big.Int) *eip1559eeEstimator {
```

Comment:
This is a bit of a discoverability problem, as WithTipMultiplier is public, but eip1559eeEstimator is private. So the generated doc will be incomplete. Any reason to not make the estimator public? We won't be able to craft an instance without using NewEIP1559FeeEstimator anyway, since it has private fields.

Alternately, you could just make the constructor look like:

```
func NewEIP1559FeeEstimator(client FeeEthClient, opts ...EIP1559FeeEstimatorOption) FeeEstimator {
  ...
}
```

and `func WithTipMultiplier(multiplier *big.Int) EIP1559FeeEstimatorOption`

Reply:
I made the `struct` public so that should be all good now.

Ad the second behavior it's an intentional design choice (maybe a bad `go` design decision, lmk). If the constructor needs to be passed a full set of options, that means that in order to create e.g. a fee estimator that increases fees if a transaction fails, the code that increases the fees needs to carry the whole information (all the options) needed to instantiate the estimator. If the object itself can be "cloned" with modifications, a custom instance can be injected and the code that increases the fees only needs to care about the fee increase instead of the customizations.